### PR TITLE
Demo Files

### PR DESCRIPTION
**Pull request by joelbarmettlerUZH**

### CODE REVIEW

File: 1_demo_howto.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+# Bounding-Box spanned by zurich tile, which is needed to determine the size of our playing area
+bottom_left_coordinate, top_right_coordinate = MapTile.bounding_box_from_maptiles_group(maptiles)
+
+# Defines 2D area using both coordinates and given resolution (meters per voxel)
+area = Area(bottom_left_coordinate, top_right_coordinate)
+
+# Use area to find out play field resolution in voxels
+[x, z] = area.dimension
+y = 100  # Set map height to 100 voxels (which is also 50 meters since we have not specified a resolution)
```
Comment (thommann) [inline 1_demo_howto.py]: it is 100m right?

File: 1_demo_howto.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+# Define environment
+environment_generator = EnvironmentGen(
+    dimensions=Coordinate4D(x, y, z, t=TIMESTEPS),
+    maptiles=maptiles,
+    map_area=area,
+    min_height=10
+)
+
+# Choose allocator, compatible payment rule and compatible mechanism
```
Comment (thommann) [inline 1_demo_howto.py]: mechanism cannot be compatible

File: 2_demo_reallocations.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+# Bounding-Box spanned by zurich tile, which is needed to determine the size of our playing area
+bottom_left_coordinate = APIWorldCoordinates(long=8.543278203323961, lat=47.36954866470366)
+top_right_coordinate = APIWorldCoordinates(long=8.545048471709597, lat=47.370620437054555)
+
+# Defines 2D area using both coordinates and given resolution (meters per voxel)
+area = Area(bottom_left_coordinate, top_right_coordinate)
+
+# Use area to find out play field resolution in voxels
+[x, z] = area.dimension
+y = 100  # Set map height to 100 voxels (which is also 50 meters since we have not specified a resolution)
```
Comment (thommann) [inline 2_demo_reallocations.py]: same

File: 2_demo_reallocations.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+# Define environment
+environment_generator = EnvironmentGen(
+    dimensions=Coordinate4D(x, y, z, t=TIMESTEPS),
+    maptiles=maptiles,
+    map_area=area,
+    min_height=0
+)
+
+# Choose allocator, compatible payment rule and compatible mechanism
```
Comment (thommann) [inline 2_demo_reallocations.py]: same

File: 2_demo_reallocations.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+environment = environment_generator.generate()
+
+# Create simulation
+simulator = Simulator(owners, mechanism, environment)
+
+# Run simulation for as long as ticks are left
+start = time_ns()
+while simulator.tick():
+    pass
+simulation_time = time_ns() - start
```
Comment (thommann) [inline 2_demo_reallocations.py]: simulation run method

File: 3_demo_collisions.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+# Bounding-Box spanned by zurich tile, which is needed to determine the size of our playing area
+bottom_left_coordinate = APIWorldCoordinates(long=8.306168879343144, lat=47.050397249055735)
+top_right_coordinate = APIWorldCoordinates(long=8.309090737025008, lat=47.05278348678034)
+
+# Defines 2D area using both coordinates and given resolution (meters per voxel)
+area = Area(bottom_left_coordinate, top_right_coordinate)
+
+# Use area to find out play field resolution in voxels
+[x, z] = area.dimension
+y = 50  # Set map height to 100 voxels (which is also 50 meters since we have not specified a resolution)
```
Comment (thommann) [inline 3_demo_collisions.py]: 50

File: 3_demo_collisions.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+# Define environment
+environment_generator = EnvironmentGen(
+    dimensions=Coordinate4D(x, y, z, t=TIMESTEPS),
+    maptiles=maptiles,
+    map_area=area,
+    min_height=0
+)
+
+# Choose allocator, compatible payment rule and compatible mechanism
```
Comment (thommann) [inline 3_demo_collisions.py]: same

File: 4_demo_fast.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+# Bounding-Box spanned by zurich tile, which is needed to determine the size of our playing area
+bottom_left_coordinate, top_right_coordinate = MapTile.bounding_box_from_maptiles_group(maptiles)
+
+# Defines 2D area using both coordinates and given resolution (meters per voxel)
+area = Area(bottom_left_coordinate, top_right_coordinate, resolution=RESOLUTION, min_height=50//RESOLUTION)
+
+# Use area to find out play field resolution in voxels
+[x, z] = area.dimension
+y = 100 // RESOLUTION  # Set map height to 100 voxels
```
Comment (thommann) [inline 4_demo_fast.py]: 100 meters

Comment (thommann) [reply 4_demo_fast.py]: 20 voxels

File: 4_demo_fast.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+# Define environment
+environment_generator = EnvironmentGen(
+    dimensions=Coordinate4D(x, y, z, t=TIMESTEPS),
+    maptiles=maptiles,
+    map_area=area,
+    min_height=50 // RESOLUTION
+)
+
+# Choose allocator, compatible payment rule and compatible mechanism
```
Comment (thommann) [inline 4_demo_fast.py]: same

File: 5_demo_weather.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
+from random import randint
+from time import time_ns
+import json
+
+from API import Area, APIWorldCoordinates, EnvironmentGen, MapTile, LongLatCoordinate, WebPathOwner, \
+    GridLocation, generate_config, generate_output
+from Demos.Priority import PriorityAllocator, PriorityPaymentRule, PriorityPathBiddingStrategy, \
+    PriorityPathValueFunction
+from Simulator import Simulator, Coordinate4D, Mechanism, StaticBlocker, DynamicBlocker
+from Simulator.Coordinates.Coordinate3D import Coordinate3D
```
Comment (thommann) [inline 5_demo_weather.py]: fix import

File: 5_demo_weather.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+# Bounding-Box spanned by zurich tile, which is needed to determine the size of our playing area
+bottom_left_coordinate, top_right_coordinate = MapTile.bounding_box_from_maptiles_group(maptiles)
+
+# Defines 2D area using both coordinates and given resolution (meters per voxel)
+area = Area(bottom_left_coordinate, top_right_coordinate)
+
+# Use area to find out play field resolution in voxels
+[x, z] = area.dimension
+y = 100  # Set map height to 100 voxels (which is also 50 meters since we have not specified a resolution)
```
Comment (thommann) [inline 5_demo_weather.py]: same

File: 5_demo_weather.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+# Define environment
+environment_generator = EnvironmentGen(
+    dimensions=Coordinate4D(x, y, z, t=TIMESTEPS),
+    maptiles=maptiles,
+    map_area=area,
+    min_height=10,
+    blockers=[*static_blockers, *dynamic_blockers],
+)
+
+# Choose allocator, compatible payment rule and compatible mechanism
```
Comment (thommann) [inline 5_demo_weather.py]: same

File: 5_demo_weather.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+# Generate config that can be interpreted by API
+simulation_config = generate_config(
+    simulator,
+    environment_generator,
+    name='Report Weather Demo Model',
+    description='This model was generated u sing the 5_demo_weather.py Script',
+    allocation_period=ALLOCATION_PERIOD,
+)
+
+# Generate simulation output that can be interpreted by API
```
Comment (thommann) [inline 5_demo_weather.py]: by the visualizer

File: API/Area.py [hunk 14]
```diff
@@ -18,7 +19,7 @@ class Area:
     and the grid size in meters
     """
 
-    def __init__(self, bottom_left_ll: "APIWorldCoordinates", top_right_ll: "APIWorldCoordinates", resolution: int,
+    def __init__(self, bottom_left_ll: Union["APIWorldCoordinates", "LongLatCoordinate"], top_right_ll: Union["APIWorldCoordinates", "LongLatCoordinate"], resolution=1,
```
Comment (thommann) [inline API/Area.py]: keep type hint on resolution

File: API/Area.py [hunk 15]
```diff
@@ -54,6 +55,17 @@ def point_to_coordinate2D(self, point: "LongLatCoordinate | APIWeightedCoordinat
         x, y = self.lon_lat_to_grid(point)
         return Coordinate2D(math.floor(x), math.floor(y))
 
+    def point_to_coordinate3D(self, point: "LongLatCoordinate | APIWeightedCoordinate", y: int):
```
Comment (thommann) [inline API/Area.py]: use union type

File: API/Generator/MapTile.py [hunk 16]
```diff
@@ -174,7 +174,7 @@ def zxy2lat(z: int, x: int, y: int) -> float:
         return float(lat)
 
     @staticmethod
-    def tiles_from_coordinates(coordinates: "APIWorldCoordinates", neighbouring_tiles: int, resolution: int) -> List[
+    def tiles_from_coordinates(coordinates: "APIWorldCoordinates", neighbouring_tiles=0, resolution=1) -> List[
```
Comment (thommann) [inline API/Generator/MapTile.py]: keep type hint

File: API/__init__.py [hunk 17]
```diff
@@ -6,3 +6,8 @@
 from .Types import APISubselection, APIWorldCoordinates
 from .config import available_allocators
 from .configGenerator import generate_config
+from .outputGenerator import generate_output
+from .WebClasses.Owners.WebPathOwner import WebPathOwner
+from .WebClasses.Owners.WebSpaceOwner import WebSpaceOwner
+from .GridLocation.GridLocation import GridLocation
+from .LongLatCoordinate import LongLatCoordinate
```
Comment (thommann) [inline API/__init__.py]: no circular imports?

File: API/configGenerator.py [hunk 18]
```diff
@@ -1,26 +1,31 @@
...
 
 from API.WebClasses.Owners.WebPathOwner import WebPathOwner
 from API.WebClasses.Owners.WebSpaceOwner import WebSpaceOwner
 from . import Area
 from .GridLocation.GridLocationType import GridLocationType
-from .LongLatCoordinate import LongLatCoordinate
-from .Types import APIBiddingStrategy, APILocations, APIMap, APIOwner, APISimulationConfig, APIWorldCoordinates
-from .WebClasses import WebBiddingStrategy
+from .Types import APIBiddingStrategy, APILocations, APIMap, APIOwner, APISimulationConfig, APIWorldCoordinates, \
+    APIWeightedCoordinate
```
Comment (thommann) [inline API/configGenerator.py]: fix imports

**Review by thommann — APPROVED** [review]
I like it very much :D

File: 2_demo_reallocations.py (hunk 4 above)
Comment (thommann) [reply 2_demo_reallocations.py]: mechanism cannot be compatible

File: 2_demo_reallocations.py (hunk 3 above)
Comment (thommann) [reply 2_demo_reallocations.py]: 100 meters

**Review by johannschwabe — APPROVED** [review]

**Pull request merged**